### ISSUE TITLE
Add parallel sorting implementation

### ISSUE THREAD

**Issue by fandrei** [dotnet/runtime#51654]
## Background and Motivation

Virtually all PCs today have more than one core, so it would be good performance wise to add a parallel sorting implementation and use that power.

## Proposed API

```
public void Sort(IComparer<T>? comparer, **ParallelOptions parallelOptions = null**)
```

## Usage Examples

```
var list = new List<int> { ... };
list.Sort(comparer, new ParallelOptions { MaxDegreeOfParallelism = 8 });
```

**Comment by Symbai** [comment]
`List<T>` is not thread safe, how is that supposed to work? And how often are list so huge that sorting can be a problem? And even when it takes 2 seconds, how often do you sort a list?

**Comment by FilipToth** [comment]
I agree with @Symbai. Lists are often small, arrays often contain a lot more items than lists. Personally, I can't recall a time when I had to sort a list.

I just don't think this API brings much value for us to implement.

**Comment by fandrei** [comment]
@Symbai same way as it's usually done - by using data synchronization primitives.

@FilipToth parallel sorting of arrays would be fine too. Or both. List uses an array as the underlying storage, anyway.

@eiriktsarpalis there are 248 calls of sorting methods in the source code of this repo. So, I assume that sorting isn't a niche application.
by my tests, an array of strings bigger than ~2000 items is big enough for parallel sorting to be faster, even when comparing a naive parallel sorting implementation to Array.Sort()
PLINQ's OrderBy  is only marginally faster than Array.Sort() - sometimes it's even slower. It's slower than an in-place parallel sorting implementation by a factor of ~1.5, and I think that it's possible to make a much faster parallel sorting implementation.

I think that there are no reasons to do it in a slow way, when it's possible to do it faster.

**Comment by Symbai** [comment]
> So, I assume that sorting isn't a niche application.

We are speaking about HUGE list's that needs to be sorted, otherwise it makes no sense to start several threads. On small lists this can be even slower than a single thread. So, how many of these 248 calls to sorting methods here are about huge lists AND are going to be called several times in a row so performance matters.

And if you are going to answer these questions you will notice, why everybody here in this room, except you, believes its a niche.

**Comment by fandrei** [comment]
@Symbai as I wrote above, bigger than ~2_000 items is already big enough. I wouldn't call this huge.
And it's easily possible to make it switch to the single threaded implementation for smaller arrays

**Comment by FilipToth** [comment]
@Symbai I never thought see a list for anything big. Yes, lists are easier to use but with arrays, you optimize (almost) every single process along the way. So, that's why I don't think this API brings much value to the user. Although, I would like to see parallel sorting of arrays but, only if you can specify whether the array will be sorted on a parallel thread or not.

**Comment by huoyaoyuan** [comment]
I did a quick benchmark using int32. For 1M items, it takes ~50ms, which may be even smaller than OS scheduling overhead. For 256M items, it takes ~21s, and PLINQ is even 2x slower.

**Comment by eiriktsarpalis** [comment]
@fandrei, would you be willing to share a benchmark with results? It would be interesting to compare `Array.Sort`, PLINQ and your in-place parallel sorting prototype with different array sizes.

**Comment by jkotas** [comment]
https://github.com/dotnet/runtime/pull/33152 discussed similar topic.

Sorting can be optimized in many ways. An optimized sorting implementation can be easily megabytes of code. It is not worth to include this amount of code in the core platforms, given how rarely sorting of large list occurs is in real world applications.

**Comment by fandrei** [comment]
@jkotas I'm very sure that it's not necessary to go to such great lengths to make a decent parallel sorting implementation.

**Comment by fandrei** [comment]
@eiriktsarpalis https://github.com/fandrei/SortingBenchmarks
```
----------------------------------------                                                                         
Int32                                                                                                            
0'002'048          Array.Sort(): 000.197          Parallel QuickSort: 000.271          Parallel OrderBy: 002.278 
0'004'096          Array.Sort(): 000.348          Parallel QuickSort: 000.238          Parallel OrderBy: 000.295 
0'008'192          Array.Sort(): 000.780          Parallel QuickSort: 000.339          Parallel OrderBy: 000.558 
0'016'384          Array.Sort(): 001.726          Parallel QuickSort: 000.634          Parallel OrderBy: 001.009 
0'032'768          Array.Sort(): 003.581          Parallel QuickSort: 001.305          Parallel OrderBy: 002.075 
0'065'536          Array.Sort(): 005.592          Parallel QuickSort: 001.660          Parallel OrderBy: 003.674 
0'131'072          Array.Sort(): 011.268          Parallel QuickSort: 003.246          Parallel OrderBy: 007.240 
0'262'144          Array.Sort(): 023.901          Parallel QuickSort: 008.476          Parallel OrderBy: 015.235 
0'524'288          Array.Sort(): 049.901          Parallel QuickSort: 016.805          Parallel OrderBy: 031.637 
1'048'576          Array.Sort(): 106.417          Parallel QuickSort: 030.783          Parallel OrderBy: 061.092 
----------------------------------------                                                                         
String                                                                                                           
0'002'048          Array.Sort(): 000.305          Parallel QuickSort: 000.397          Parallel OrderBy: 000.561 
0'004'096          Array.Sort(): 000.635          Parallel QuickSort: 000.324          Parallel OrderBy: 000.578 
0'008'192          Array.Sort(): 001.418          Parallel QuickSort: 000.553          Parallel OrderBy: 001.036 
0'016'384          Array.Sort(): 003.174          Parallel QuickSort: 001.170          Parallel OrderBy: 001.956 
0'032'768          Array.Sort(): 006.870          Parallel QuickSort: 002.213          Parallel OrderBy: 004.175 
0'065'536          Array.Sort(): 013.209          Parallel QuickSort: 003.421          Parallel OrderBy: 007.423 
0'131'072          Array.Sort(): 033.054          Parallel QuickSort: 009.971          Parallel OrderBy: 021.638 
0'262'144          Array.Sort(): 079.006          Parallel QuickSort: 031.544          Parallel OrderBy: 055.077 
0'524'288          Array.Sort(): 182.694          Parallel QuickSort: 053.379          Parallel OrderBy: 134.959 
1'048'576          Array.Sort(): 434.727          Parallel QuickSort: 133.190          Parallel OrderBy: 299.790 
----------------------------------------                                                                         
```

**Comment by Symbai** [comment]
Tested a list with all strings of 80 char (that is the maximum in your benchmark) with either 2,000 (that is what you said should be super slow and worth to run parallel) and 10,000 what I believe is already a niche in real world. And I sorted them with `StringComparer.Ordinal` because you did as well. Result:

|   Method |     N |      Mean |     Error |    StdDev |
|--------- |------ |----------:|----------:|----------:|
| SortTest |  2000 |  95.49 us |  0.800 us |  0.856 us |
| SortTest | 10000 | 701.36 us | 11.956 us | 11.183 us |

95us.... that's 0,000094 seconds. I can just repeat myself here, even if you can make that faster, it would be a micro optimization. Sorting is only called one time, unlike other methods in a row. Its like you want to make your car faster but optimizing the time you need to refill your fuel tank at the shell station.


Here is the full benchmark:

```C#
using System;
using System.Collections.Generic;
using System.Linq;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Jobs;
using BenchmarkDotNet.Running;
using ConsoleApp6;

namespace ConsoleApp6
{
    [SimpleJob(RuntimeMoniker.NetCoreApp50)]
    public class BenchmarkClass
    {
        [Params(2000, 10000)]
        public int N;

        private List<string> stringList;

        [GlobalSetup]
        public void Setup()
        {
            stringList = new List<string>(N);
            for (int i = 0; i < N; i++)
            {
                stringList.Add(GenerateRandomString());
            }
        }

        [Benchmark]
        public int SortTest()
        {
            stringList.Sort(StringComparer.Ordinal);
            return stringList.Count;
        }


        public static string GenerateRandomString()
        {
            return new string(Enumerable.Repeat(chars, 80)
                .Select(s => s[random.Next(s.Length)]).ToArray());
        }
        static readonly Random random = new Random();
        const string chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
    }
}

class Program
{

    static void Main(string[] args)
    {
        var summary = BenchmarkRunner.Run<BenchmarkClass>();
    }
}
```

**Comment by FilipToth** [comment]
After reviewing the benchmarks, I suppose we could implement parallel sorting for arrays. And, it would probably be a good idea to introduce an optional parameter to the already existing sort method specifying the thread. And, I suppose we could aim at releasing this in .NET 6.0.

**Comment by fandrei** [comment]
@Symbai your benchmark is incorrect. Out of 212994 repeats that it runs, 212992 are executed on already sorted data.

**Comment by fbrosseau** [comment]
Parallel algorithms are a science and they probably belong to special-purpose nugets rather than the framework. 

Adding this to something as fundamental as list/array would most certainly confuse beginners (why wouldn't you want parallelism, right? ;)) into using this when inappropriate and actually get worse code.

**Comment by Symbai** [comment]
> @Symbai your benchmark is incorrect. Out of 212994 repeats that it runs, 212992 are executed on already sorted data.

Fine then:

|   Method |     N |       Mean |    Error |   StdDev |
|--------- |------ |-----------:|---------:|---------:|
| SortTest |  2000 |   262.0 us |  3.34 us |  2.79 us |
| SortTest | 10000 | 1,471.0 us | 11.24 us | 13.81 us |

```C#
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Jobs;
using BenchmarkDotNet.Running;
using ConsoleApp6;

namespace ConsoleApp6
{
    [SimpleJob(RuntimeMoniker.NetCoreApp50)]
    public class BenchmarkClass
    {
        [Params(2000, 10_000)]
        public int N;

        private List<string> stringList;

        [IterationCleanup]
        [GlobalSetup]
        public void Setup()
        {
            stringList = new List<string>(N);
            for (int i = 0; i < N; i++)
            {
                stringList.Add(GenerateRandomString());
            }
        }

        [Benchmark]
        public int SortTest()
        {
            stringList.Sort(StringComparer.Ordinal);
            return stringList.Count;
        }


        public static string GenerateRandomString()
        {
            return new string(Enumerable.Repeat(chars, 80)
                .Select(s => s[random.Next(s.Length)]).ToArray());
        }
        static readonly Random random = new Random();
        const string chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
    }
}

class Program
{

    static void Main(string[] args)
    {
        var summary = BenchmarkRunner.Run<BenchmarkClass>();
    }
}
```

0.00147 seconds for 10,000 strings with 80 char length. I still don't see why this needs micro optimization.

**Comment by FilipToth** [comment]
> Parallel algorithms are a science and they probably belong to special-purpose nugets rather than the framework.
> 
> Adding this to something as fundamental as list/array would most certainly confuse beginners (why wouldn't you want parallelism, right? ;)) into using this when inappropriate and actually get worse code.

We'll just add parallelism as something optional. The sort methods can still be used without specifying the parallel thread or other parameters. It could cause errors when not used properly, I get your point.

**Comment by rickbrew** [comment]
btw I have a straightforward implementation of this that I use in Paint.NET (v4.2.16+) https://github.com/paintdotnet/PaintDotNet.Quantization/blob/d1f276d733cc13ae818a7a1ccfd34143222d369e/PaintDotNet/Collections/ListUtil.cs#L186

Some discussion here https://twitter.com/rickbrewPDN/status/1388871279994818560

I see 6x faster on a Ryzen 5950X. Probably can be tweaked/optimized to be even faster, I haven't toyed with it much.

**Comment by WhiteBlackGoose** [comment]
> (why wouldn't you want parallelism, right? ;))

@fbrosseau By this logic why wouldn't you want `Parallel.For` for loops? Isn't it also the same kind of confusion for beginners?

**Comment by fandrei** [comment]
@Symbai no, not fine. You should respect other people's time and better check your data.

**Comment by FilipToth** [comment]
> @fbrosseau By this logic why wouldn't you want `Parallel.For` for loops? Isn't it also the same kind of confusion for beginners?

How would you incorporate parallelism into for loops? We would have to change the whole structure of how we write them, that would be a breaking change.

**Comment by Symbai** [comment]
> @Symbai no, not fine. You should respect other people's time and better check your data.

What's wrong with my data this time?

**Comment by fandrei** [comment]
@Symbai the length of strings isn't same as in my benchmark. And I never said that sorting of arrays of 2000 items must be "super-slow".

**Comment by fandrei** [comment]
@FilipToth I think he means that using a parallel on/off param in the Sort() method is not more "science" and not more confusing than using Parallel.For, which was added to .NET long time ago.

**Comment by Symbai** [comment]
> @Symbai the length of strings isn't same as in my benchmark. And I never said that sorting of arrays of 2000 items must be "super-slow".

Its 80 length char and you said 10'000 is slow. Its not, it takes very less than 1 seconds. Again , whats wrong? I'm more than happy to provide a benchmark with you data.

**Comment by FilipToth** [comment]
@fandrei Oh, right. I was thinking about adjusting the for loop syntax to include parallelism which would obviously be a terrible idea. But, I still think this would beneficial.

**Comment by fandrei** [comment]
@Symbai it's not 80 chars in my bench, and I said that it's slow**er** than parallel sort (and it actually is)

**Comment by fandrei** [comment]
@FilipToth by the way, it seems that .NET already has an implementation of parallel merge sort that's used internally. https://github.com/dotnet/runtime/blob/main/src/libraries/System.Linq.Parallel/src/System/Linq/Parallel/Utils/Sorting.cs

**Comment by FilipToth** [comment]
@fandrei Great! Now to decide if we want to expose that, any thoughts?
